[RFC] Port to meson

okaestne opened the pull request:
* Update Packaging
	* updated deps
	* updated rules
	* debhelper-compat 12
	* add gir-package for introspection typelib
* Remove many obsolete files
	* autotools related files (Makefile.am)
	* nowhere referenced:
		* `cut-n-paste/`
		* `xviewer-toolbar.xml`
		* `xviewer.convert`
	* `menu` and its icon (`gnome-xviewer.xpm`), we use a .desktop file
	* `debian/patches`, already applied
	* empty files (`HACKING`, `NEWS`, `TODO`)
* (POTENTIALLY) BREAKING CHANGES
	* changed paths:
		* `/usr/lib/xviewer` => `/usr/lib/<arch>/xviewer`
			* ~~breaks plugins, if their packaging won't be changed as well~~
			* plugins will stay in `$prefix/lib/xviewer/plugins`
			* xviewer's source code has been fixed
		* `/usr/share/{appdata => metainfo}/xviewer.appdata.xml`
		* moved UI-resources to common `src/resources` directory
		* `/usr/include/{xviewer-3.0/xviewer => xviewer}/`
* other changes
	* add `.gitignore`
	* add a more detailed `README.md`
	* removed unneeded header in `xviewer-print-image-setup.c` (caused compile error)
	* many small other things

Any feedback is welcome :) At most regarding the library path. My plan would be to port the plugins to meson as well and by doing also change the library path.
Another thing: what about giving the library a so-version and moving it out of it's directory directly into `/usr/lib/x86_64-linux-gnu/`?

The dependencies (debian/control) could also need a few more tweaks.

TODO

- [x] check handling of optional deps (XMP, JPEG), regarding macros etc.
- [x] ~check jpegutils handling (drop v6 code?)~ libjpeg is only available in version 6 on debian
- [x] fix LDME build

okaestne commented:
Thanks for you feedback, @Fantu.

* I wasn't aware that -dbg packages are discouraged by now. Dropped it like the 'Guide for Debian Maintainers' recommends so that the -dbgsym package is built automatically.
* dh_dwz has been disabled because it says 'DWARF compression not beneficial - old size 671696 new size 685433' otherwise.
* Don't think it's necessary to explicitly set the `-Wl,-O1` flag, so I left it out

Will look into the LMDE build issue asap

okaestne commented:
Does the appdata.xml offer any value? If I understand the docs correctly, the .desktop file can provide most information (and currently even more than the appdata.xml).

Also maybe a question for @Fantu again: whats the best way to fix the lintian error  `build-depends-indep-without-arch-indep`? Adding the -doc packages to the build deps? These are afaik only "required" for links in the generated docs..

leigh123linux commented:
> Does the appdata.xml offer any value? If I understand the docs correctly, the .desktop file can provide most information (and currently even more than the appdata.xml).
> 

appdata.xml is used by gnome-software to display packages, the .desktop file doesn't do that.

Fantu commented:
@okaestne about "build-depends-indep-without-arch-indep" is because don't have any package with arch indipendent files (for example xviewer-common), is an error on lintian because put arch-indip. files in arch-indip. package help to decrease repository size (in debian with many archs supported is useful) but split require attention, also other changes, unfortunately recently I don't have time to check well and help you with this packaging improvement, unfortunately I don't even have enough to keep well the debian packages of which I am maintainer/co-maintainer. with just a quick look at the commits as I have done so far, some issue may be missed

okaestne commented:
rebased and updated version of meson project

eli-schwartz commented on data/meson.build:
```diff
@@ -0,0 +1,36 @@
+    'appdata',
+    input: 'xviewer.appdata.xml.in',
+    output: 'xviewer.appdata.xml',
+    command: [intltool_merge, '-x', '-u', po_dir, '@INPUT@', '@OUTPUT@'],
```
Can't https://mesonbuild.com/i18n-module.html#i18nmerge_file be used for this?

okaestne replied:
Updated, thanks!

eli-schwartz commented on data/meson.build:
```diff
@@ -0,0 +1,36 @@
+    'desktop',
+    input: 'xviewer.desktop.in.in',
+    output: 'xviewer.desktop',
+    command: [intltool_merge, '-d', '-u', po_dir, '@INPUT@', '@OUTPUT@'],
```
https://mesonbuild.com/i18n-module.html#i18nmerge_file again

eli-schwartz commented on meson.build:
```diff
@@ -0,0 +1,169 @@
+cc = meson.get_compiler('c')
+
+# on some systems we need to find the math lib to make sure it builds
+math = cc.find_library('m')
```
On some systems, meaning it should be `required: false` so that systems which don't have it, still work.

Any system will either have it, or not need it -- it is part of the base system and distributed with libc.

https://mesonbuild.com/howtox.html#add-math-library-lm-portably

okaestne replied:
Thanks for the explanation, this makes sense to me now :) Fixed.

eli-schwartz commented on meson_options.txt:
```diff
@@ -0,0 +1,48 @@
+option(
+    'enable_exempi',
+    type: 'boolean',
```
Please use https://mesonbuild.com/Build-options.html#features for this.

okaestne replied:
thanks for the advice

eli-schwartz commented on meson.build:
```diff
@@ -0,0 +1,169 @@
+# (optional) XMP support
+if enable_exempi
+    exempi = dependency('exempi-2.0', version: '>= 1.99.5')
+    xviewer_conf.set10('HAVE_EXEMPI', true)
```
... and use them here.

```meson
exempi = dependency('exempi-2.0', version: '>= 1.99.5', required: get_option('exempi'))
if exempi.found()
    xviewer_conf.set10('HAVE_EXEMPI', true)
endif
```

This lets you specify enabled/disabled or *auto*, and, in the disabled case, it actually logs a helpful status message:

```
Dependency exempi skipped: feature exempi disabled
```

eli-schwartz commented on src/meson.build:
```diff
@@ -0,0 +1,218 @@
+
+if enable_exif
+    xviewer_deps += exif
+    libxviewer_sources += 'xviewer-exif-util.c'
```
```meson
if exif.found()
    libxviewer_sources += 'xviewer-exif-util.c'
    public_headers += 'xviewer-exif-util.h'
endif
# this will just be a no-op if it isn't found
xviewer_deps += exif
```

You could add all the xviewer_deps in one list rather than piecemeal.

This is all possible because of using feature options.

eli-schwartz reviewed (changes requested):
A bit difficult to be sure I saw all the meson changes, since the removal of the autotools build system is not factored out into a distinct commit...

okaestne commented:
Thank you very much @eli-schwartz for reviewing! I incorporated your tips and suggestions.
If you want to, I can squash the new commits (e.g. with the first one in the series) and rebase on master.

Btw, I left the intltool dependency in debian/control, just because the pot-file generation is still using it. Maybe this should be migrated to gettext (?) as well. But this pull request is already quite comprehensive.

eli-schwartz commented on meson.build:
```diff
@@ -152,17 +147,19 @@ endif
+summary_msg += '  deprecated_warnings: ' + enable_deprecation_warnings.to_string() + '\n'
+summary_msg += '  docs               : ' + enable_docs.to_string() + '\n'
+summary_msg += 'Features:\n'
+summary_msg += '  exempi             : ' + get_option('exempi').enabled().to_string() + '\n'
```
I don't think it is particularly useful to just print the value of the feature option. What about taking this opportunity to, instead, print whether the dependency itself was found?

okaestne replied:
How about that:
```
feat_fmt = '  @0@             : enabled: @1@, found: @2@\n'
features = {
    # display string (w/padding), option, dependency (found?)
    'exempi': [get_option('exempi'), exempi],
    'exif  ': [get_option('exif'), exif],
    'jpeg  ': [get_option('jpeg'), jpeg],
    'lcms  ': [get_option('lcms'), lcms],
    'rsvg  ': [get_option('rsvg'), rsvg]
}

foreach feat_str, feat_val : features
    summary_msg += feat_fmt.format(feat_str, feat_val[0].enabled(), feat_val[1].found())
endforeach
```
result:
```
Features:
  exempi             : enabled: true, found: true
  exif               : enabled: true, found: true
  jpeg               : enabled: true, found: true
  lcms               : enabled: true, found: true
  rsvg               : enabled: true, found: true
```
But Meson also prints whether the deps have been found before the summary itself..
```
Run-time dependency exempi-2.0 found: YES 2.5.1
Run-time dependency libexif found: YES 0.6.21
Run-time dependency libjpeg found: YES 2.0.3
Checking if "libjpeg version is 8 or greater" with dependency libjpeg compiles: YES 
Run-time dependency lcms2 found: YES 2.9
Run-time dependency librsvg-2.0 found: YES 2.48.9
```

So I don't know how helpful this output is at all..

eli-schwartz replied:
Some people do like a final summary, and if you look at it as a summary of components to be built then it makes a bit of sense.

My real concern here is that just printing the value of the *option* is downright misleading, for example when the build was configured using `-Dexempi=auto` and the dependency gets found, this will print "Features: exempi: false" even though the software is built with support for exempi.

The argument in favor of summaries is that they aggregate data from earlier in the log, and list all the optional dependencies in one group alongside their configured statuses.

okaestne replied:
Ah I think I now understand what you mean. Setting options to `auto` in meson.options results in `enabled: false` in the summary print-out, because it's `'auto'` and thus not `'enabled'`. So we should instead check `dep.found()` to _print_ whether the feature is "enabled" in the build. Correct?
Also, I now think it might be better to set all features in `meson.options` to `auto`, but in `debian/rules` to `enabled`

okaestne replied:
changed as written in my previous comment

eli-schwartz commented:
I am personally a fan of squashing "fixup" style commits, yes.

As for intltool, I don't think it is needed in the debian packaging considering that pot-file generation is done during development, not as part of building a debian package. Although porting it too, to the core gettext tools if possible, probably cannot hurt. :)


And for general context on the change, there is https://wiki.gnome.org/MigratingFromIntltoolToGettext

okaestne commented:
I just rebased the PR on master, squashed some commits and fixed the feature summary as discussed with @eli-schwartz . 

I would just like to know what your views are on the proposed changes:
* dbg -> dbgsym package and
* move of the plugins directory (which is currently reverted by b125d9a to compatibility with the xviewer-plugins repo)

mtwebster commented:
- debug package: I'd prefer to leave it as xviewer-dbg_xxx.deb. I know it's obsolete but that's how it is currently, and there's no reason to go thru obsoleting/replacing a package. I'm also not a fan of having different type of packages generated by the same build (deb vs ddeb).
- plugin directory - it's probably ok to move, but at the same time as we update xviewer-plugins (and we'll need to make sure to depend on minimum versions to support the change). For now leave it where the plugins currently go.

Everything looks fine, but can you add back some brief instruction to build it in the readme (or add the INSTALL file back, either is fine).  Just 'on mint/ubuntu/debian use dpkg-buildpackage,  anywhere else use meson builddir etc....'

Thanks

mtwebster commented:
Thanks a lot!
